src/Modules/Iterators.jl

```diff
@@ -0,0 +1,89 @@
+### Iteration over monomials in modules of a certain degree
+struct AllModuleMonomials{ModuleType<:FreeMod}
+  F::ModuleType
+  d::Int
+
+  function AllModuleMonomials(F::FreeMod{T}, d::Int) where {T <: MPolyDecRingElem}
+    is_graded(F) || error("module must be graded")
+    S = base_ring(F)
+    is_standard_graded(S) || error("iterator implemented only for the standard graded case")
+    return new{typeof(F)}(F, d)
+  end
+end
+
+underlying_module(amm::AllModuleMonomials) = amm.F
+degree(amm::AllModuleMonomials) = amm.d
+
+function all_monomials(F::FreeMod{T}, d::Int) where {T<:MPolyDecRingElem}
+  return AllModuleMonomials(F, d)
+end
+
+Base.eltype(amm::AllModuleMonomials{T}) where {T} = elem_type(T)
+
+function Base.length(amm::AllModuleMonomials)
+  F = underlying_module(amm)
+  r = rank(F)
+  R = base_ring(F)
+  d = degree(amm)
+  result = 0
+  for i in 1:r
+    d_loc = d - Int(degree(F[i])[1])
+    d_loc < 0 && continue
+    result = result + length(all_monomials(R, d_loc))
+  end
+  return result
+end
+  
+function Base.iterate(amm::AllModuleMonomials, state::Nothing = nothing)
+  i = 1
+  F = underlying_module(amm)
+  d = degree(amm)
+  R = base_ring(F)
+
+  d_loc = d - Int(degree(F[i])[1])
+  while i < ngens(F) && d_loc < 0
+    i = i+1
+    d_loc = d - Int(degree(F[i])[1])
+  end
+  
+  i == ngens(F) && d_loc < 0 && return nothing
+
+  mon_it = all_monomials(R, d_loc)
+  res = iterate(mon_it, nothing)
+  res === nothing && i == ngens(F) && return nothing
```

Comment:
Can `res === nothing` happen? `mon_it` is an iterator over a component of non-negative degree, so there should always be at least one monomial, no?

Reply:
I don't know. Probably not. But the line doesn't hurt.

Reply:
But what does happen if `res === nothing` and `i != ngens(F)`, so the `return nothing` is not reached?
In the next line, you assign `x, s = res` which throws an error.
Handling this case properly (increase `i` etc.) might be possible, but doesn't seem to be worth the effort.
```suggestion
  @assert res !== nothing
```
Maybe this would be clearer then?

Reply:
Yes, you might be right. But I don't want to interrupt the tests again now. Let's first see whether they pass in principle. In either case, the line doesn't seem to be breaking things. Worst case it's confusing to the reader.